doc/performance/0002-http-vs-rcp-client-performance-analysis.md

```diff
@@ -0,0 +1,94 @@
+4. Log the time difference to the console
+
+This were done manually simply because measuring the performance after the JSON-RPC patch requires a bit more effort, which involves `playwright testing`. This would be an interesting task to be done in the future if we decide that we need more detailed information regarding the performance post-JSON-RPC. For now, manual tests would be sufficient.
+
+The three functions that were tested are:
+
+1. tspClient.experimentOutputs(): Open 4 traces. I used the 4 traces in the 103 sample traces of traviz-labs. Then reload the extension to trigger the function. Make sure all trace tabs are closed. This has the least amount of data sent to the Theia BE with just 1 parameter.
+
+2. tspClient.fetchXYTree(): Open a histogram view. This requires a few more parameters sent to the Theia BE compare to the prior test case.
+
+3. tspClient.fetchTimeGraphStates(): Open a Flame Chart. I used the bad-trace sample that were used in the timeline-chart performance analysis. This requires the most amount of data to be sent to the Theia BE.
+
+## Results
+
+### Pre-JSON-RPC-patch
+
+|experimentOutputs() (ms)(*)| fetchXYTree() (ms)| fetchTimeGraphStates() (ms)
+|---|---|---|
+|472.5|90.7|100.8|
+|301|98.4|108.2|
+|465.3|95.7|139.2|
+|865.9|91.6|126.7|
+|905.1|103.6|201.8|
+
+On average we have:
+
+|experimentOutputs() (ms)(*)| fetchXYTree() (ms)| fetchTimeGraphStates() (ms)
+|---|---|---|
+|601.6|96.0|135.34|
+
+(*): For the case of fetchTimeGraphStates(), the number varies quite a bit. If we run more test attempts, sometimes we have outliers that are very small (24-35ms). However, it looks like most of the time, the range is 400-600ms.
+
+### Post-JSON-RPC-patch
+
+|experimentOutputs() (ms)(**)| fetchXYTree() (ms)| fetchTimeGraphStates() (ms)
+|---|---|---|
+|503.4|92.5 |292.8|
+|160.1|95.7|142.9|
+|116.5|97.9|128.5|
+|159.2|100.5|133.8|
+|175.9|91|123.8|
+
+On average we have:
+
+|experimentOutputs() (ms)(**)| fetchXYTree() (ms)| fetchTimeGraphStates() (ms)
+|---|---|---|
+|223.0|94.3|164.5|
+
+(**): Similar to (\*) sometimes we have outliers too (around 26-35ms).
+
+## Conclusion
+
+The results are not consistent for each different test cases. On average:
+
+1. For experimentOutputs(), the JSON-RCP update improves the speed of communication.
+2. For fetchXYTree(), the JSON-RCP update does not make much difference in the speed of communication.
+3. For fetchTimeGraphStates(), the JSON-RCP update takes more time compared to the HTTP direct implementation.
+
+One of the possible explanation is the difference in the amount of data that needs to be sent to the back end. This is true for the case of the `fetchXYTree()` and the `fetchTimeGraphStates()` tests. Since they both serialize a good amount of data, the performance either stays the same - in the case of `fetchXYTree()` where we have small amount of data - or becomes worse - in the case of `fetchTimeGraphStates()` where we have a decent amount of data.
+

```

Comment:
A missing test case might be, to measure the impact on a trailing yet tiny payload request, preceded by a heavy one. Meaning, is the trailing (tiny) request starved in any way by the preceding elephant. We could compare the outcome of this test before versus after json-rpc. This was recently proposed by @MatthewKhouzam IIRC.

So I think he mentioned the tiny request being a trace annotations/markers one or so. -Preceded by some truckload data fetch act. The latter shouldn't starve or prevent the former from showing, IIUC. If this is an issue before changing to json-rpc, then measuring with the latter in turn might become irrelevant. The question could otherwise be, is the json-rpc wrapper (towards the back-ended client use) adding latency to the trailing/tiny request?

Reply:
@MatthewKhouzam since you are back, do you think this is a required test case?